src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3WebtransportSession.cs

```diff
@@ -0,0 +1,184 @@
+        internal const string VersionHeaderPrefix = $"{SecPrefix}draft";
+        internal const string CurrentSuppportedVersion = $"{VersionHeaderPrefix}02";
+
+        internal Http3WebtransportSession(QuicStream connectStream, Http3WebtransportManager manager)
+        {
+            _WebtransportManager = manager;
+            _connectStream = connectStream;
+            _ = _connectStream.WritesClosed.ContinueWith(async t =>
+            {
+                await DisposeAsync().ConfigureAwait(false);
+
+            }, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Current);
+
+        }
+
+        /// <summary>
+        /// Creates a webtransport session by creating a webtransport connect request, sending it to <see cref="Uri">uri</see>.
+        /// </summary>
+        public static async ValueTask<Http3WebtransportSession?> ConnectAsync(Uri uri, HttpMessageInvoker? invoker, CancellationToken cancellationToken)
+        {
+            if (invoker is null)
+            {
+                s_defaultInvoker ??= new HttpMessageInvoker(new HttpClientHandler());
+                invoker = s_defaultInvoker;
+            }
+
+            Http3WebtransportSession? webtransportSession;
+            try
+            {
+                HttpRequestMessage request;
+                request = new HttpRequestMessage(HttpMethod.Connect, uri) { Version = HttpVersion.Version30, VersionPolicy = HttpVersionPolicy.RequestVersionExact };
+                request.Headers.Protocol = WebTransportProtocolValue;
+                Task<HttpResponseMessage> sendTask = invoker is HttpClient client ? client.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken) : invoker.SendAsync(request, cancellationToken);
+                HttpResponseMessage response = await sendTask.ConfigureAwait(false);
+                WebtransportHttpContent connectedWebtransSessionContent = (WebtransportHttpContent)response.Content;
+                webtransportSession = connectedWebtransSessionContent.webtransportSession;
+                if (!response.IsSuccessStatusCode || !response.Headers.Contains(Http3WebtransportSession.VersionHeaderPrefix))
+                {
+                    await webtransportSession.AbortIncomingSessionWebtransportStreams((long)Http3ErrorCode.WebtransportBufferedStreamRejected).ConfigureAwait(false);
+                    await webtransportSession.DisposeAsync().ConfigureAwait(false);
+                    throw new(SR.net_webtransport_server_rejected);
+                }
+            }
+            catch (HttpRequestException ex)
+            {
+                throw ex;
+            }
+
+            return webtransportSession;
+        }
+
+        /// <summary>
+        /// Takes the next incoming <see cref="QuicStream">quic stream from the server</see>.
+        /// </summary>
+        public async ValueTask<QuicStream?> GetIncomingWTStreamFromServerAsync()
```

Comment:
what will happen if server never initiates any stream from their side and client will call this method?

Reply:
@CarnaViire  It will wait until the channel is closed aka the session is finished.
I will create a try-catch inside and I will return null in case the channel is closed
Also I will add a function TryGetIncomingWTStreamFromServer which will return a bool and the stream in case it is true

Reply:
Solved adding CancellationToken and created a new test to atest it